File: day4/part2/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	file, err := os.Open("../day4.txt")
	if err != nil {
		fmt.Println("Failed to open file")
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	count := 0

	for i, s := range lines[1 : len(lines)-1] {
		i++
		for j, c := range s[1 : len(s)-1] {
			j++
			if c == 'A' {
				topLeft := lines[i-1][j-1]
				topRight := lines[i+1][j-1]
				bottomLeft := lines[i-1][j+1]
				bottomRight := lines[i+1][j+1]
				if topRight == 'M' && topLeft == 'M' && bottomRight == 'S' && bottomLeft == 'S' {
					count++
				} else if topRight == 'M' && topLeft == 'S' && bottomRight == 'M' && bottomLeft == 'S' {
					count++
				} else if topRight == 'S' && topLeft == 'S' && bottomRight == 'M' && bottomLeft == 'M' {
					count++
				} else if topRight == 'S' && topLeft == 'M' && bottomRight == 'S' && bottomLeft == 'M' {
					count++
				}
			}
		}
	}
	fmt.Println(count)
}

func dfs(i int, j int, matrix []string, word string, direction []int) int {
	if word == "" {
		return 1
	}
	newX := i + direction[0]
	newY := j + direction[1]
	if newX < 0 || newX >= len(matrix) || newY < 0 || newY >= len(matrix[0]) {
		return 0
	}

	if matrix[newX][newY] == word[0] {
		return dfs(newX, newY, matrix, word[1:], direction)
	}
	return 0
}

File: day6/part2.py
def getDirection(direction):
    if direction == (-1, 0):
        return (0, 1)
    elif direction == (0, 1):
        return (1, 0)
    elif direction == (1, 0):
        return (0, -1)
    elif direction == (0, -1):
        return (-1, 0)


f = open('day6.txt', 'r')

matrix = []
start = None
i = 0
for line in f:
    matrix.append([])
    j = 0
    for c in line:
        if c == '.':
            matrix[i].append(True)
        elif c == '#':
            matrix[i].append(False)
        elif c == '^':
            matrix[i].append(True)
            start = (i, j)
        j += 1
    i += 1

count = 0
for i in range(len(matrix)):
    for j in range(len(matrix[0])):
        if matrix[i][j] and (i, j) != start:
            x, y = start
            direction = (-1, 0)
            visited = set()
            matrix[i][j] = False
            while True:
                newX, newY = x + direction[0], y + direction[1]
                if (newX, newY, direction[0], direction[1]) in visited:
                    count += 1
                    break
                if not (newX >= 0 and newX < len(matrix) and newY >= 0 and newY < len(matrix[0])):
                    break
                if matrix[newX][newY]:
                    x = newX
                    y = newY
                    visited.add((x, y, direction[0], direction[1]))
                else:
                    direction = getDirection(direction)
            matrix[i][j] = True


print(count)
